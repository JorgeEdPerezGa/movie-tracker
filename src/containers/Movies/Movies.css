.movie-list {
  justify-content: center;
  display: flex;
  background: #0c0c0c;
  flex-wrap: wrap;
  padding: 30px; }
