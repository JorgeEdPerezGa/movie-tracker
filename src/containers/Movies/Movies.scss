.movie-list {
  display: flex;
  background: #0c0c0c;
  flex-wrap: wrap;
  padding: 30px;
}
