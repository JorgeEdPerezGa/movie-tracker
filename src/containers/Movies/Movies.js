import React, { Component } from 'react';
import { connect } from 'react-redux';
import MovieCard from '../../components/MovieCard/MovieCard';
import { addFavorite, removeFavorite, updateMovies } from '../../actions/';
import { postFavorite, deleteFavorite } from '../../helper';
import './Movies.css';

export class Movies extends Component {

  displayMovies = () => {
    const { movies } = this.props;
    return movies.map((movie, idx) => {
      console.log(movie)
      return (
        <MovieCard key={movie.movie_id + idx} movie={movie} onFavorite={this.handleFavorites}/>
      );
    });
  }

  handleFavorites = (movie) => {
    if (!this.props.user.name) {
      return this.props.history.push('/login');
    }

    const duplicated = this.props.favorites.some(fav => movie.title === fav.title);

    const favMovie = {...movie, favorite: !movie.favorite};

    duplicated ? this.removeFavMovie(favMovie) :this.addFavMovie(favMovie);

    this.props.updateMovies(favMovie);
  }

  addFavMovie = (movie) => {
    this.props.addFavorite(movie);
    postFavorite(movie, this.props.user);
  }

  removeFavMovie = (movie) => {
    this.props.removeFavorite(movie);
    deleteFavorite(this.props.user, movie)
  }

  render() {
    return (
      <div>
        {this.displayMovies()}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  movies: state.movies,
  user: state.user,
  favorites: state.favorites
});

const mapDispatchToProps = (dispatch) => ({
  addFavorite: movie => dispatch(addFavorite(movie)),
  removeFavorite: movie => dispatch(removeFavorite(movie)),
  updateMovies: movie => dispatch(updateMovies(movie))
})

export default connect(mapStateToProps, mapDispatchToProps)(Movies);
