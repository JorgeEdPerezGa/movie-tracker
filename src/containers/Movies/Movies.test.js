import React from 'react';
import { shallow } from 'enzyme';
import { Movies } from './Movies';
import { createMemoryHistory } from 'history';
jest.mock('../../helper');

describe('Movies', () => {

  let mockFavorites;
  let mockMovies;
  let mockUser;
  let mockAddFavorite;
  let mockRemoveFavorite;
  let mockUpdateMovies;
  let renderedComponent;
  let mockMovie;

  beforeAll(() => {
    mockMovie = {
      backdrop_path: "/askg3SMvhqEl4OL52YuvdtY40Yb.jpg",
      movie_id: 354912,
      overview: "Despite his family’s baffling generations-old ban on music",
      poster_path: "/eKi8dIrr8voobbaGzDpe8w0PVbC.jpg",
      release_date: "2017-10-27",
      title: "Coco",
      vote_average: 7.7,
      favorite: false
    };

    mockMovies = [
      {
        backdrop_path: "/askg3SMvhqEl4OL52YuvdtY40Yb.jpg",
        movie_id: 354912,
        overview: "Despite his family’s baffling generations-old ban on music",
        poster_path: "/eKi8dIrr8voobbaGzDpe8w0PVbC.jpg",
        release_date: "2017-10-27",
        title: "Coco",
        vote_average: 7.7,
        favorite: false
      },
      {
        backdrop_path: "/def.jpg",
        movie_id: 999,
        overview: "Despite his family’s baffling generations-old ban on music",
        poster_path: "/abs.jpg",
        release_date: "2017-10-27",
        title: "Thor",
        vote_average: 7.2,
        favorite: true
      }
    ];

    mockUser = {
      email: "fake@123.com",
      id: 2,
      name: "ted",
      password: "pw123"
    };
  });

  beforeEach(() => {
    mockRemoveFavorite = jest.fn();
    mockAddFavorite = jest.fn();
    mockUpdateMovies = jest.fn();
    window.fetch = jest.fn();

    mockFavorites = [
      {
        backdrop_path: "/def.jpg",
        movie_id: 999,
        overview: "Despite his family’s baffling generations-old ban on music",
        poster_path: "/abs.jpg",
        release_date: "2017-10-27",
        title: "Thor",
        vote_average: 7.2,
        favorite: true
      }
    ];

    renderedComponent = shallow(
      <Movies
        favorites={mockFavorites}
        movies={mockMovies}
        user={mockUser}
        addFavorite={mockAddFavorite}
        removeFavorite={mockRemoveFavorite}
        updateMovies={mockUpdateMovies}
      />
    );
  });

  it('should match snapshot', () => {
    expect(renderedComponent).toMatchSnapshot();
  });

  it('when removeFavMovie is called removeFavorite and deleteFavorite should be called with expected params', () => {
    expect(mockRemoveFavorite).not.toHaveBeenCalled();
    expect(window.fetch).not.toHaveBeenCalled();

    renderedComponent.instance().removeFavMovie(mockMovie);

    expect(mockRemoveFavorite).toHaveBeenCalledWith(mockMovie);
    expect(window.fetch).toHaveBeenCalledWith( mockMovie, mockUser);
  });

  it('when addFavMovie is called addFavorite and postFavorite should be called with expected params', () => {
    expect(mockAddFavorite).not.toHaveBeenCalled();
    expect(window.fetch).not.toHaveBeenCalled();

    renderedComponent.instance().addFavMovie(mockMovie);

    expect(mockAddFavorite).toHaveBeenCalledWith(mockMovie);
    expect(window.fetch).toHaveBeenCalledWith(mockMovie, mockUser);
  });

  it.skip('when user favorites a card and is not logged in, page redirects to login', () => {
    const history = createMemoryHistory('/movies');
    const renderedComponent = shallow(
      <Movies
        favorites={mockFavorites}
        movies={mockMovies}
        user={{}}
        addFavorite={mockAddFavorite}
        removeFavorite={mockRemoveFavorite}
        updateMovies={mockUpdateMovies}
        history={history}
      />
    );

    renderedComponent.instance().handleFavorites(mockMovie);
    expect(renderedComponent).toMatchSnapshot();
  });

  it('when user favorites a card and it is already in favorites, it should call removeFavMovie', () => {
    const expected = {...mockFavorites[0], favorite: false};

    expect(mockRemoveFavorite).not.toHaveBeenCalled();
    renderedComponent.instance().handleFavorites(mockFavorites[0]);
    expect(mockRemoveFavorite).toHaveBeenCalledWith(expected);
  });

  it('when user favorites a card and it is not in Favorites, it should call addFavMovie', () => {
    const expected = {...mockMovie, favorite: true};

    expect(mockAddFavorite).not.toHaveBeenCalled();
    renderedComponent.instance().handleFavorites(mockMovie);
    expect(mockAddFavorite).toHaveBeenCalledWith(expected);
  });

  it('when user favorites a card it should update Movies', () => {
    const expected = {...mockMovie, favorite: true};

    expect(mockUpdateMovies).not.toHaveBeenCalled();
    renderedComponent.instance().handleFavorites(mockMovie);
    expect(mockUpdateMovies).toHaveBeenCalledWith(expected);
  });
});
