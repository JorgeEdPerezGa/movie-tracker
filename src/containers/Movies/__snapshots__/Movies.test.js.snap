// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movies when user favorites a card and is not logged in, page redirects to login 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Movies
    addFavorite={[Function]}
    favorites={
        Array [
            Object {
              "backdrop_path": "/def.jpg",
              "favorite": true,
              "movie_id": 999,
              "overview": "Despite his family’s baffling generations-old ban on music",
              "poster_path": "/abs.jpg",
              "release_date": "2017-10-27",
              "title": "Thor",
              "vote_average": 7.2,
            },
          ]
    }
    history={
        Object {
            "action": "PUSH",
            "block": [Function],
            "canGo": [Function],
            "createHref": [Function],
            "entries": Array [
              Object {
                "hash": "",
                "key": "zqx71f",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              Object {
                "hash": "",
                "key": "mttltz",
                "pathname": "/login",
                "search": "",
                "state": undefined,
              },
            ],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "index": 1,
            "length": 2,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "key": "mttltz",
              "pathname": "/login",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
    }
    movies={
        Array [
            Object {
              "backdrop_path": "/askg3SMvhqEl4OL52YuvdtY40Yb.jpg",
              "favorite": false,
              "movie_id": 354912,
              "overview": "Despite his family’s baffling generations-old ban on music",
              "poster_path": "/eKi8dIrr8voobbaGzDpe8w0PVbC.jpg",
              "release_date": "2017-10-27",
              "title": "Coco",
              "vote_average": 7.7,
            },
            Object {
              "backdrop_path": "/def.jpg",
              "favorite": true,
              "movie_id": 999,
              "overview": "Despite his family’s baffling generations-old ban on music",
              "poster_path": "/abs.jpg",
              "release_date": "2017-10-27",
              "title": "Thor",
              "vote_average": 7.2,
            },
          ]
    }
    removeFavorite={[Function]}
    updateMovies={[Function]}
    user={Object {}}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <MovieCard
          movie={
                    Object {
                              "backdrop_path": "/askg3SMvhqEl4OL52YuvdtY40Yb.jpg",
                              "favorite": false,
                              "movie_id": 354912,
                              "overview": "Despite his family’s baffling generations-old ban on music",
                              "poster_path": "/eKi8dIrr8voobbaGzDpe8w0PVbC.jpg",
                              "release_date": "2017-10-27",
                              "title": "Coco",
                              "vote_average": 7.7,
                            }
          }
          onFavorite={[Function]}
/>,
        <MovieCard
          movie={
                    Object {
                              "backdrop_path": "/def.jpg",
                              "favorite": true,
                              "movie_id": 999,
                              "overview": "Despite his family’s baffling generations-old ban on music",
                              "poster_path": "/abs.jpg",
                              "release_date": "2017-10-27",
                              "title": "Thor",
                              "vote_average": 7.2,
                            }
          }
          onFavorite={[Function]}
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "354912",
        "nodeType": "function",
        "props": Object {
          "movie": Object {
            "backdrop_path": "/askg3SMvhqEl4OL52YuvdtY40Yb.jpg",
            "favorite": false,
            "movie_id": 354912,
            "overview": "Despite his family’s baffling generations-old ban on music",
            "poster_path": "/eKi8dIrr8voobbaGzDpe8w0PVbC.jpg",
            "release_date": "2017-10-27",
            "title": "Coco",
            "vote_average": 7.7,
          },
          "onFavorite": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "1000",
        "nodeType": "function",
        "props": Object {
          "movie": Object {
            "backdrop_path": "/def.jpg",
            "favorite": true,
            "movie_id": 999,
            "overview": "Despite his family’s baffling generations-old ban on music",
            "poster_path": "/abs.jpg",
            "release_date": "2017-10-27",
            "title": "Thor",
            "vote_average": 7.2,
          },
          "onFavorite": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <MovieCard
            movie={
                        Object {
                                    "backdrop_path": "/askg3SMvhqEl4OL52YuvdtY40Yb.jpg",
                                    "favorite": false,
                                    "movie_id": 354912,
                                    "overview": "Despite his family’s baffling generations-old ban on music",
                                    "poster_path": "/eKi8dIrr8voobbaGzDpe8w0PVbC.jpg",
                                    "release_date": "2017-10-27",
                                    "title": "Coco",
                                    "vote_average": 7.7,
                                  }
            }
            onFavorite={[Function]}
/>,
          <MovieCard
            movie={
                        Object {
                                    "backdrop_path": "/def.jpg",
                                    "favorite": true,
                                    "movie_id": 999,
                                    "overview": "Despite his family’s baffling generations-old ban on music",
                                    "poster_path": "/abs.jpg",
                                    "release_date": "2017-10-27",
                                    "title": "Thor",
                                    "vote_average": 7.2,
                                  }
            }
            onFavorite={[Function]}
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "354912",
          "nodeType": "function",
          "props": Object {
            "movie": Object {
              "backdrop_path": "/askg3SMvhqEl4OL52YuvdtY40Yb.jpg",
              "favorite": false,
              "movie_id": 354912,
              "overview": "Despite his family’s baffling generations-old ban on music",
              "poster_path": "/eKi8dIrr8voobbaGzDpe8w0PVbC.jpg",
              "release_date": "2017-10-27",
              "title": "Coco",
              "vote_average": 7.7,
            },
            "onFavorite": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "1000",
          "nodeType": "function",
          "props": Object {
            "movie": Object {
              "backdrop_path": "/def.jpg",
              "favorite": true,
              "movie_id": 999,
              "overview": "Despite his family’s baffling generations-old ban on music",
              "poster_path": "/abs.jpg",
              "release_date": "2017-10-27",
              "title": "Thor",
              "vote_average": 7.2,
            },
            "onFavorite": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
